test/test_cuda.py

```diff
@@ -425,6 +425,79 @@ def tmp(self):
 
 
 class TestCuda(TestCase):
+
+    def test_memory_allocated_stats(self):
+        torch.cuda.empty_cache()
+        m0 = last_m = torch.cuda.memory_allocated()
+        max_m = torch.cuda.max_memory_allocated()
+
+        def assert_change(comp=1):
+            # comp > 0: increased
+            # comp = 0: equal
+            # comp < 0: decreased
+            nonlocal last_m, max_m
+            new_m = torch.cuda.memory_allocated()
+            new_max_m = torch.cuda.max_memory_allocated()
+            if comp > 0:
+                self.assertGreater(new_m, last_m)
+            elif comp < 0:
+                self.assertLess(new_m, last_m)
+            else:
+                self.assertEqual(new_m, last_m)
+            self.assertLessEqual(new_m, new_max_m)
+            self.assertGreaterEqual(new_max_m, max_m)
+            last_m = new_m
+            max_m = new_max_m
+
+        tensors1 = [torch.randn(1).cuda(), torch.randn(10, 20).cuda(), torch.randn(200, 300, 2000).cuda()]
+        m1 = torch.cuda.memory_allocated()
+        assert_change(1)
+
+        tensors2 = []
+        N = 30
+
+        for i in range(1, int(N / 2) + 1):
+            # small ones
+            tensors2.append(torch.randn(i, i * 3).cuda())
+        assert_change(1)
+
+        for i in range(1, N - len(tensors2) + 1):
+            # large ones
+            tensors2.append(torch.randn(i * 7, i * 9, i * 11).cuda())
+        assert_change(1)
+
+        tensors2.append(torch.randn(0, 0, 0).cuda())
+        assert_change(0)
+
+        permute = []
+        for i in torch.randperm(len(tensors2)):
+            permute.append(tensors2[i])
+            assert_change(0)
+
+        del tensors2
+        assert_change(0)
+        tensors2 = permute
+        assert_change(0)
+        del permute
+        assert_change(0)
+
+        for i in range(int(N / 2)):
+            x = tensors2[i].numel()
+            del tensors2[i]
+            assert_change(-x)
```

Comment:
should this just be `-1`? `-x` looks as if you were trying to assert that it dropped exactly by `x` (which isn't true - you're missing a factor of `element_size`)

Reply:
This is used to deal with the empty tensor added at line 469 :) If the selected tensor is empty, it asserts that the memory usage stays the same.